react-components/src/components/RuleBasedOutputs/RuleBasedOutputsSelector.tsx

```diff
@@ -1,40 +1,87 @@
+import {
+  CogniteCadModel,
+  TreeIndexNodeCollection,
+  type CogniteModel,
+  NumericRange,
+  type NodeAppearance
+} from '@cognite/reveal';
 import { useAllMappedEquipmentAssetMappings } from '../..';
 import { Color } from 'three';
-import { type RuleOutputSet } from './types';
+import { type AssetStylingGroupAndStyleIndex } from './types';
 import { generateRuleBasedOutputs } from './utils';
-import { type FdmPropertyType } from '../Reveal3DResources/types';
 import { use3dModels } from '../../hooks/use3dModels';
+import { type AssetMapping3D } from '@cognite/sdk/dist/src';
+import { filterUndefined } from '../../utilities/filterUndefined';
+import { type AssetStylingGroup, type FdmPropertyType } from '../Reveal3DResources/types';
 
 export type ColorOverlayProps = {
-  ruleSet: RuleOutputSet | Record<string, any> | FdmPropertyType<Record<string, any>> | undefined;
+  ruleSet: Record<string, any> | FdmPropertyType<Record<string, any>> | undefined;
+  onRuleSetChanged?: (currentStylings: AssetStylingGroupAndStyleIndex[] | undefined) => void;
 };
 
-export function RuleBasedOutputsSelector({ ruleSet }: ColorOverlayProps): ReactElement | undefined {
+export function RuleBasedOutputsSelector({
+  ruleSet,
+  onRuleSetChanged
+}: ColorOverlayProps): ReactElement | undefined {
   const models = use3dModels();
 
+  const [stylingGroups, setStylingsGroups] = useState<AssetStylingGroupAndStyleIndex[]>();
+
   const {
     data: assetMappings,
     isFetching,
     hasNextPage,
     fetchNextPage
   } = useAllMappedEquipmentAssetMappings(models);
 
+  const applyBasedNodeStyling = (
+    model: CogniteModel,
+    assetMappings: AssetMapping3D[]
+  ): AssetStylingGroupAndStyleIndex | undefined => {
+    if (!(model instanceof CogniteCadModel)) {
+      return;
+    }
+
+    const baseNodeStyling = new TreeIndexNodeCollection();
+    assetMappings?.forEach((assetMapping) => {
+      const range = new NumericRange(assetMapping.treeIndex, assetMapping.subtreeSize);
+      baseNodeStyling.getIndexSet().addRange(range);
+    });
+    const nodeAppearance: NodeAppearance = {
+      color: new Color('#efefef')
+    };
+    const assetStylingGroup: AssetStylingGroup = {
+      assetIds: assetMappings.map((node) => node.assetId),
+      style: { cad: nodeAppearance }
+    };
+
+    const baseStylingGroup: AssetStylingGroupAndStyleIndex = {
+      styleIndex: baseNodeStyling,
+      assetStylingGroup
+    };
+
+    return baseStylingGroup;
+  };
+
   useEffect(() => {
     if (!isFetching && hasNextPage === true) {
       void fetchNextPage();
     }
   }, [isFetching, hasNextPage, fetchNextPage]);
 
   useEffect(() => {
-    if (assetMappings === undefined || isFetching) return;
+    if (onRuleSetChanged !== undefined) onRuleSetChanged(stylingGroups);
+  }, [stylingGroups]);
 
-    clearAllModelStyling(models);
+  useEffect(() => {
+    if (assetMappings === undefined || models === undefined || isFetching) return;
+
+    setStylingsGroups([]);
```

Comment:
Can you use `EMPTY_ARRAY` instead of `[]` here? It makes it so that the reference stays the same on subsequent runs of this statement. Same goes for other setter-calls using an empty array